### PR TITLE
feat(pytest-bdd): add support for 6.1+

### LINKED ISSUE #4492: support pytest-bdd>=6.1
### Which version of dd-trace-py are you using?

Effected versions: `ddtrace>=1.3`

### Which version of pip are you using?

Not applicable

### Which libraries and their versions are you using?

``pytest-bdd>=6.1``

### How can we reproduce your problem?

Run ddtrace pytest-bdd tests with  ``pytest-bdd>=6.1``. 

First Failure
- Sample Error: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/23378/workflows/7d24fef6-27cf-4bc9-959a-b8df7cbd64c2/jobs/1590909
- Cause: pytest_bdd.__version__ was removed.

Second Failure
- Sample Error: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/23379/workflows/b7bbdb4e-e0df-4f68-8c2b-2b6a320d9c9f/jobs/1591008 
- Cause:  ddtrace uses accesses [step_func._pytest_bdd_parsers](https://github.com/DataDog/dd-trace-py/blob/v1.6.0/ddtrace/contrib/pytest_bdd/plugin.py#L80) to parse pytest arguments. This attribute was [removed in v6.1](https://github.com/pytest-dev/pytest-bdd/pull/534/files)

### What is the result that you get?

CI failures

### What is the result that you expected?

No CI Failures

### PR DESCRIPTION
**Pull request by romainkomorndatadog**
Enable support for `pytest-bdd >= 6.1` by:
* Using `importlib_metadata` or `importlib.metadata` as necessary to gather package version
* Switching to collection `step_func_args` after steps finish (or after they error)
* Introducing a "safe" JSONification of `step_func_args` that catches possible `TypeError` exceptions and returns the error in a `json.dumps()`-safe dict

Compatibility with `pytest-bdd < 4` is maintained by moving the old codepath to a utility function and using it whenever it looks like we have the old-style functions available. Technically, this maintains compatibility with `< 6.1` because that's when the latter function was removed but, based on testing, `step_func_args` is populated as expected in `>= 4.0` and our test suites pass for Python `3.6+`.

The backwards-compatible code is necessary because older versions of `pytest-bdd` include fixtures in the `step_func_args` passed to the `pytest_bdd_after_step` and `pytest_bdd_step_error`, breaking `json.dumps()`.

Also in this change:
* Updating dependencies for testing environments
* Returning `pytest-bdd` versions in `riotfile` to their state pre #4492 .

Fixes #4492 .

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](../docs/contributing.rst#release-branch-maintenance))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](../docs/contributing.rst#release-branch-maintenance)

### CODE REVIEW

**Comment by romainkomorndatadog** [comment]
`pytest-bdd` support was added in version `1.3`, and https://github.com/DataDog/dd-trace-py/pull/4486 introduced the pinning of `pytest-bdd` in `1.7` (if I'm not mistaken) , so I'm not sure how far back this should be backported.

File: ddtrace/contrib/pytest_bdd/plugin.py [hunk 1]
```diff
@@ -25,6 +25,39 @@ def _store_span(item, span):
     setattr(item, "_datadog_span", span)
 
 
+def _extract_step_func_args(step, step_func, step_func_args):
```
Comment (romainkomorndatadog) [inline ddtrace/contrib/pytest_bdd/plugin.py:28]: I'm thinking of adding a unit test for this, even if it's fairly directly tested by existing tests.

Comment (gnufede) [reply ddtrace/contrib/pytest_bdd/plugin.py:28]: It's your call, if it's already covered I think it's enough

**Review by gnufede — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

**Comment by majorgreys** [comment]
While beyond adding support for 6.1.x, do we want to address the risk you flag with json serialization?

**Comment by romainkomorndatadog** [comment]
Ignore the `@pytest.mark.romain` decorators for now, sorry, using it for local testing but I wanted to get a Circle CI run done.

**Review by majorgreys — APPROVED** [review]

**Pull request merged**